src/ser/serializer.rs

```diff
@@ -0,0 +1,422 @@
+//! Custom MQTT message serializer
+//!
+//! # Design
+//! This serializer handles serializing MQTT packets.
+//!
+//! This serializer does _not_ assume MQTT packet format. It will encode data linearly into a
+//! buffer and converts all data types to big-endian byte notation.
+//!
+//! The serializer reserves the first number of bytes for the MQTT fixed header, which is filled
+//! out after the rest of the packet has been serialized.
+//!
+//! # Limitations
+//! It is the responsibility of the user to handle prefixing necessary lengths and types on any
+//! MQTT-specified datatypes, such as "Properties", "Binary Data", and "UTF-8 Encoded Strings".
+//!
+//! # Supported Data Types
+//!
+//! Basic data types are supported, including:
+//! * Signed & Unsigned integers
+//! * Booleans
+//! * Strings
+//! * Bytes
+//! * Options (Some is encoded as the contained contents, None is not encoded as any data)
+//! * Sequences
+//! * Tuples
+//! * Structs
+//!
+//! Other types are explicitly not implemented and there is no plan to implement them.
+use crate::message_types::MessageType;
+use crate::varint::VarintBuffer;
+use bit_field::BitField;
+use serde::Serialize;
+use varint_rs::VarintWriter;
+
+/// The maximum size of the MQTT fixed header in bytes. This accounts for the header byte and the
+/// maximum variable integer length.
+const MAX_FIXED_HEADER_SIZE: usize = 5;
+
+/// Errors that result from the serialization process
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Error {
+    /// The provided memory buffer did not have enough space to serialize into.
+    InsufficientMemory,
+
+    /// The requested feature will not be implemented.
+    WontImplement,
```

Comment:
I suspect in general this variant isn't needed and code should `unimplemented!()`
Same for `de`.